### PR TITLE
feat: display progress during prologue

### PR DESCRIPTION
**Pull request by dariuszd21**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Introduce more detailed logging during environment preparation:
- possibility to pass standard output and error from upstream during ExecutionContext context initialization steps
- wire passed streams to track host package installation

Addresses: #739

### CODE REVIEW

**Review by tigarmo — COMMENTED** [review]
Thanks! I think this PR needs a "companion PR" in craft-application to showcase the changes that the applications need to make to use this new API. It would also be great (and I should say, essential) to be able to see what the end-result looks like in say snapcraft. Maybe a screen recording? Otherwise it's hard to see what the impact of this change is.

File: craft_parts/packages/yum.py [hunk 1]
```diff
@@ -115,13 +115,16 @@ def download_packages(cls, package_names: List[str]) -> None:
         """
         raise NotImplementedError("Functionality not yet provided by YUMRepository.")
 
+    # TODO (Dariusz): Update also this signature and calls
```
[the remark this answers is not in the document]
Comment (dariuszd21) [reply craft_parts/packages/yum.py:118]: Actually, that's a leftover.
I marked this place to not forget to update it in here and when I updated it I forgot to remove the comment.

**Comment by dariuszd21** [comment]
> Thanks! I think this PR needs a "companion PR" in craft-application to showcase the changes that the applications need to make to use this new API. It would also be great (and I should say, essential) to be able to see what the end-result looks like in say snapcraft. Maybe a screen recording? Otherwise it's hard to see what the impact of this change is.

For the companion PR, I've created draft one in snapcraft to showcase the difference.
The `stage-package` part application does not have to do any changes as it already gathers .info logs for streaming.
Streaming `apt install` output requires application to wire stdout/stderr to the whole execution context rather than just `.run()`. In such sense .run() wiring could be removed in non-backward compatible version as we already have an ouput in an object.
https://github.com/canonical/snapcraft/pull/4831

I'll try to record and attach how changes looks in real-world scenario

**Comment by dariuszd21** [comment]
Here is the video on how does it work on snapcraft "draft" branch.
[snapcraft_wired_apt_install.webm](https://github.com/canonical/craft-parts/assets/9116056/c1e36c22-3935-48ec-b701-9e4e0f1800dd)

**Review by mr-cal — APPROVED** [review]
This looks good and works for me.

When we fix https://github.com/canonical/craft-providers/issues/250 in craft-providers, this will be a good reference.

**Comment by dariuszd21** [comment]
As discussed, we'll not introduce wiring for apt output so this is canceled.

**Pull request closed without merging**